src/primitives/poseidon.rs

```diff
@@ -0,0 +1,361 @@
+use std::iter;
+use std::marker::PhantomData;
+
+use halo2::arithmetic::FieldExt;
+
+pub(crate) mod grain;
+pub(crate) mod mds;
+
+mod nullifier;
+pub use nullifier::OrchardNullifier;
+
+use grain::SboxType;
+
+/// A specification for a Poseidon permutation.
+pub trait Spec<F: FieldExt> {
+    /// The type used to hold permutation state, or equivalent-length constant values.
+    ///
+    /// This must be an array of length [`Spec::arity`], that defaults to all-zeroes.
+    type State: Default + AsRef<[F]> + AsMut<[F]>;
+
+    /// The type used to hold duplex sponge state.
+    ///
+    /// This must be an array of length equal to the rate of the duplex sponge (allowing
+    /// for a capacity consistent with this specification's security level), that defaults
+    /// to `[None; RATE]`.
+    type Rate: Default + AsRef<[Option<F>]> + AsMut<[Option<F>]>;
+
+    /// The arity of this specification.
+    fn arity() -> usize;
```

Comment:
```suggestion
    /// The width of this specification.
    fn width() -> usize;
```
The [Poseidon paper](https://eprint.iacr.org/2019/458.pdf) uses "width" for this, not "arity". It uses "arity" only in the context of Merkle tree arity (which is not equal to the width, when Poseidon is used for a Merkle tree).

Reply:
Yeah, I renamed this in the circuit PR, but forgot to do so here.